Utils Bindings

codeboy5 opened the pull request:
# Pull Request

## Description
I have added a binding for `mean` function. Also added a test in `test_util.py`. @chinmayshah99 is this okay?

## Affected Dependencies
List any dependencies that are required for this change.

## Type of Change
Please mark options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation (non-breaking change which adds documentation)
- [x] Improvement (non-breaking change that improves the performance or reliability of existing functionality)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How has this been tested?
`pytest tests`

## Checklist
- [x] I did follow the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md)
- [x] I did follow the [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have added tests for my changes

## Additional Context
Includes bindings for #27

codeboy5 committed:
utils bindings

codeboy5 committed:
style fix

chinmayshah99 commented on src/bindings/PyDP/algorithms/util.cpp:
```diff
@@ -14,4 +15,6 @@ void init_algorithms_util(py::module& m) {
   util.def("get_next_power_of_two", &dp::GetNextPowerOfTwo);
   util.def("qnorm", &dp::Qnorm);
+  util.def("mean",&dp::Mean<double>);
+  // util.def("hello", &dp:: )
```
I think a helper function is needed to covert the python list to a vector. Is this code tested using pytest?

codeboy5 replied:
yes I have added a test, you can check that in this commit

codeboy5 replied:
@chinmayshah99 should I proceed to change these to have a wrapper function which will allow us to specify the type ?

chinmayshah99 replied:
Yeah, please go ahead.

codeboy5 committed:
more bindings and tests

codeboy5 committed:
more bindings and tests

codeboy5 committed:
bindings for add,subtract and sqaure

codeboy5 committed:
style fix and TODO added

chinmayshah99 commented on src/bindings/PyDP/algorithms/util.cpp:
```diff
@@ -14,4 +15,29 @@ void init_algorithms_util(py::module& m) {
+  util.def("vector_to_string", &dp::VectorToString<double>);
+  util.def("round_to_nearest_multiple", &dp::RoundToNearestMultiple);
+  // TODO: Throw an exception in case the function returns false.
+  util.def("safe_add", [](int64_t i, int64_t j) {
```
How to change datatype when we want to use floats?

codeboy5 replied:
If we do something like `(double i, double j){}` it would automatically use that if passed a float.

chinmayshah99 replied:
So the way we are supposed to use this is: 
`safe_add(number1, number2, 'int'/'float')`

codeboy5 replied:
How would we do something like this in plain c++ ?

codeboy5 committed:
throwing runtime error during invalid operations

codeboy5 commented:
> # Pull Request
> ## Description
> I have added a binding for `mean` function. Also added a test in `test_util.py`. @chinmayshah99 is this okay?
> 
> ## Affected Dependencies
> List any dependencies that are required for this change.
> 
> ## Type of Change
> Please mark options that are relevant.
> 
> * [ ]  Bug fix (non-breaking change which fixes an issue)
> * [ ]  Documentation (non-breaking change which adds documentation)
> * [x]  Improvement (non-breaking change that improves the performance or reliability of existing functionality)
> * [x]  New feature (non-breaking change which adds functionality)
> * [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
> 
> ## How has this been tested?
> `pytest tests`
> 
> ## Checklist
> * [x]  I did follow the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md)
> * [x]  I did follow the [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
> * [ ]  I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
> * [x]  I have added tests for my changes
> 
> ## Additional Context
> Includes bindings for #27

@chinmayshah99 I have added exceptions for the invalid operations.

chinmayshah99 commented on src/bindings/PyDP/algorithms/util.cpp:
```diff
@@ -14,4 +15,32 @@ void init_algorithms_util(py::module& m) {
+  util.def("vector_filter", &dp::VectorFilter<double>);
+  util.def("vector_to_string", &dp::VectorToString<double>);
+  util.def("round_to_nearest_multiple", &dp::RoundToNearestMultiple);
+  // TODO: Throw an exception in case the function returns false.
```
Is this still a TODO?

chinmayshah99 replied:
@codeboy5 can you remove this comment if it's done already or is it left as a future exercise?

chinmayshah99 reviewed (dismissed):
LGTM!

chinmayshah99 commented:
@codeboy5 if it's no longer WIP, can we remove the tag from the PR title?

codeboy5 commented:
> @codeboy5 if it's no longer WIP, can we remove the tag from the PR title?

Yes I think this is done from my side!

codeboy5 committed:
removed TODO
